src/gevent/pool.py

```diff
@@ -714,19 +717,39 @@ def free_count(self):
             return 1
         return max(0, self.size - len(self))
 
-    def add(self, greenlet):
+    def add(self, greenlet, blocking=True, timeout=None):
         """
-        Begin tracking the given greenlet, blocking until space is available.
+        Begin tracking the given greenlet, possibly blocking until space is
+        available.
+
+        :keyword bool blocking: If True (the default), this function will block
+        until the pool has space or a timeout occurs.
+        :keyword float timeout: The maximum number of seconds this method will block.
+
+        :return: True if the greenlet was added; False if a Timeout occured or
+        if blocking is True and the pool is full.
 
         .. seealso:: :meth:`Group.add`
         """
-        self._semaphore.acquire()
+        try:
+            # NOTE: The docs for Semaphore.acquire indicate that it may raise a Timeout rather
+            # than return False under some circumstances, though I'm not sure exactly what those
+            # circumstances are.
+            was_acquired = self._semaphore.acquire(blocking=blocking, timeout=timeout)
+        except Timeout:
+            was_acquired = False
+
+        if not was_acquired:
+            return False
+
         try:
             Group.add(self, greenlet)
         except:
             self._semaphore.release()
             raise
 
+        return True
```

Comment:
I know Semaphore has the pattern of returning True/False to indicate that it was acquired or that the timeout elapsed. I'm not sure that's the right thing to do in general, though. Return values are easy to ignore or miss. 

Semaphore gets away with it because of how low-level speed critical it is (it's compiled with Cython, and not having the possibility of raising exceptions at all makes a big difference), but I don't think our constraints are that tight here. What if this did what `subprocess.Popen.communicate` does and raises an exception if the semaphore can't be acquired (`class PoolFull(Queue.Full)`)?

Reply:
Yep, I agree that it's cleaner to have `Pool.add` raise a `Timeout` rather than return True/False. Will modify.

Reply:
I don't think `Timeout` is correct. If `blocking` is `False`, we could return immediately. I think we need a new exception, which is why I suggested `class PoolFull(Queue.Full)`. `Timeout` is a `BaseException`, not an `Exception` (the better to bubble up to the caller who is expecting it), but I think we want this exception to be caught by a generic `except Exception:` block.

Reply:
Yep, makes sense. Couple of questions:

1. The name `Full` makes more sense to me than `PoolFull`, since it's already qualified by the module name (`pool.Full`).
2. It feels unnatural to have `pool.Full` (or whatever we decide to call it`) derive from `Queue.Full`, since classes `Pool` and `Queue` have no relationship. Does it confer some benefit to associate the two? If not, then I'd just as soon do `class Full(Exception)`.

@jamadden

Reply:
OK, so here's what I was thinking. I'm not committed to it, but it seemed to make a certain amount of sense.

For inheritance:

A stdlib Queue (which shares the same API as a `gevent.queue.Queue`) and a `Pool` have similar APIs. They share `full`, `join` and possibly some other methods I'm not remembering.  One of the supported use cases of a Queue is as a task queue, where daemon threads are blocking and waiting on tasks to finish; it provides the `task_done` callback method for this purpose. Conceptually, that's what a `Pool` is, and this PR makes them even more similar by adding blocking and timeout parameters to `Pool.add`, in much the same way that `Queue.put` has them. `Queue.put` raises `Queue.Full` when it fails given those parameters. 

It wouldn't be hard to transition from a stdlib Queue + threads as a task queue to a gevent `Pool`. 

If that's under a layer or two of abstraction, it would be nice if we had exception compatibility too so that higher level code could continue catching the old `Queue.Full` exception even if the implementation changes. 

Is that a real use case? I don't really know. But it didn't seem to hurt much to try to support it.

On naming:

It's not uncommon to use both a Queue and a Pool at the same time. You might have 1000 websites to scrape but want to limit the concurrency to 20 at a time. So you put all the websites in a Queue and use a Pool to spin up 20 workers who each draw from that Queue. It would be nice to be able to import  classes from both modules without having to manually worry about renaming conflicting classes.

Plus, on Python 2, only the local name of an exception is printed in a traceback, so if this exception ever got logged or bubbled all the way to the top, it would be good if the name was unambiguous without the module.

Reply:
Sounds good, I find that convincing. Making the changes now.

Reply:
FYI I've coded `gevent.pool` to import `Full` from `gevent.queue` rather than from the stdlib, since that module already handles the Python2/Python3 juggling. Let me know how the implementation looks to you. Will commit once the tests complete.

Btw, while I was in there, I noticed that `gevent.queue.__all__` does not contain `Full` or `Empty`, whereas the stdlib `Queue.__all__` does. I'm going to submit a separate PR where I'll propose bringing gevent's `queue.__all__` to parity with the stdlib's.